_posts/2018-12-13-arms-final-report.md

```diff
@@ -0,0 +1,151 @@
+
+This post outlines the development of ARMS (Automated Robotic Markup Syntax) in the Fall 2018 semester at Missouri State University. ARMS is a TOML syntax that is parsed by a Python program called Armature. The purpose of ARMS is to consolidate the specification of robotic parts and joints into one file, while Armature takes that file and outputs the selected target simulation code. For example, currently Armature can output both SDF and C++ code to generate the same robot in both Gazebo and ODE simulation environments respectively.
+
+![Lines of code comparison.](/assets/2018-12-13-arms-final-report/lines_of_code.png)
+
+As the figure above illustrates, one file of about 60 lines of ARMS code can output hundreds of lines of working code in other languages. This allows the user to more quickly experiment and tweak the design of a robot, while giving them the source code of that robot to allow them to further fine tune the details to their liking. If the user is working in one simulation environment and it is not working quite how they expect it should, they can run Armature with their ARMS file and instantly have the new environments code without having to write any new code themselves.
+
+ This post will discuss the syntax of ARMS that we chose and how we came to that decision. Second, this post will discuss the features of ARMS that were developed in the Fall 2018 semester. Lastly, this post will discuss development that could take place in the future. An example 'car' ARMS code is included at the bottom of this post for reference while reading and to use as an example ARMS file. 
+
+# Syntax of ARMS
+
+The first and most important decision that was made about the design of ARMS was how we would like to write and parse it. It was initially decided that we were going to write our own syntax and parser so we could have maximum control over the syntax. Although, it was quickly found that writing a syntax, parser, and creating functionality all in one semester was not realistic.
+
+It was decided that having a working proof-of-concept was the highest priority to pursue first, so we decided to not write an original language and use an existing one that is well defined and supported. After some research, it was decided that TOML was the language that would support our needs and be the fasted to get started with.
+
+The key functionality that made TOML attractive was the concept of an 'Array of Tables' which provides us with syntax like this:
+
+```
+[[sphere]]
+name = "wheel1"
+radius = 0.5
+relative_position = [1.0, 1.0, 0.0]
+
+[[sphere]]
+name = "wheel2"
+radius = 0.5
+relative_position = [1.0, -1.0, 0.0]
+```
+
+This code illustrates how TOML creates arrays of dictionaries. These two '[[sphere]]' objects will be stored in one array as dictionaries with the values below them being key-value pairs. This allows Armature to organize all objects of the same type in a convenient manner and handle all of the same type of object simultaneously.
```

Comment:
Maybe show the equivalent Python dictionary object?

Reply:
Done